### PR TITLE
hold namespaces briefly before processing deletion

### PR DESCRIPTION
**Pull request by liggitt**
possible fix for #36891

in HA scenarios (either HA apiserver or HA etcd), it is possible for deletion of resources from namespace cleanup to race with creation of objects in the terminating namespace

HA master timeline:
1. "delete namespace n" API call goes to apiserver 1, deletion timestamp is set in etcd
2. namespace controller observes namespace deletion, starts cleaning up resources, lists deployments
3. "create deployment d" API call goes to apiserver 2, gets persisted to etcd
4. apiserver 2 observes namespace deletion, stops allowing new objects to be created
5. namespace controller finishes deleting listed deployments, deletes namespace

HA etcd timeline:
1. "create deployment d" API call goes to apiserver, gets persisted to etcd
2. "delete namespace n" API call goes to apiserver, deletion timestamp is set in etcd
3. namespace controller observes namespace deletion, starts cleaning up resources, lists deployments
4. list call goes to non-leader etcd member that hasn't observed the new deployment or the deleted namespace yet
5. namespace controller finishes deleting the listed deployments, deletes namespace

In both cases, simply waiting to clean up the namespace (either for etcd members to observe objects created at the last second in the namespace, or for other apiservers to observe the namespace move to terminating phase and disallow additional creations) resolves the issue

Possible other fixes:
* do a second sweep of objects before deleting the namespace
* have the namespace controller check for and clean up objects in namespaces that no longer exist
* ...?

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37431)

### CODE REVIEW

File: pkg/controller/namespace/namespace_controller.go [hunk 1]
```diff
@@ -132,7 +132,9 @@ func (nm *NamespaceController) enqueueNamespace(obj interface{}) {
 		glog.Errorf("Couldn't get key for object %+v: %v", obj, err)
 		return
 	}
-	nm.queue.Add(key)
+	// delay processing namespace deletions long enough for the lifecycle admission
+	// plugins to also observe the deletion and stop new objects from being created
+	nm.queue.AddAfter(key, 5*time.Second)
```
Comment (liggitt) [inline pkg/controller/namespace/namespace_controller.go]: @derekwaynecarr I think we need this in both HA master and HA etcd scenarios, including when the master is behind a load balancer and the endpoints only list a single IP. Not sure about the duration

Comment (kargakis) [inline pkg/controller/namespace/namespace_controller.go]: This change will affect all actions, not just deletions, right? On a different note, it seems we need a way to identify if admission has run. Would a new generation/observedGeneration pair make sense?

Comment (liggitt) [inline pkg/controller/namespace/namespace_controller.go]: > On a different note, it seems we need a way to identify if admission has run

can you clarify what you mean by that? admission exists on every API server in HA scenarios, admission only runs on a deletion on the API server that handles the delete request, and anything persisted in etcd has run through admission already

Comment (kargakis) [inline pkg/controller/namespace/namespace_controller.go]: Why admission doesn't reject new objects for namespaces that have the deletion timestamp set? Wouldn't that solve the issue with a HA master? Regarding a HA etcd, is the namespace controller getting all its watch events from the master but can try to list in a different etcd member?

Comment (liggitt) [inline pkg/controller/namespace/namespace_controller.go]: > Why admission doesn't reject new objects for namespaces that have the deletion timestamp set?

It does, but in an HA apiserver scenario, knowledge that a namespace has a deletionTimestamp set takes non-zero time to propagate, so apiserver B can allow a create after a namespace was deleted via apiserver A

**Comment by derekwaynecarr** [comment]
For the HA API server case, we could have every server register itself as a finalizer on all namespaces, we could then have the "kubernetes" finalizer only run when all other finalizers are removed, which means we would know that every API server in the deployment has observed the watch event.  The problem with that solution is that it makes scale up and scale down of the API server a more challenging process since we need to dynamically add/remove a finalizer as part of server start-up / shutdown.  This also doesn't fix the HA etcd scenario, so having a large engineering effort like that seems problematic.

I think a sleep for now is a reasonable remedy for this release.

File: pkg/controller/namespace/namespace_controller.go (hunk 1 above)
Comment (liggitt) [inline pkg/controller/namespace/namespace_controller.go]: > Regarding a HA etcd, is the namespace controller getting all its watch events from the master but can try to list in a different etcd member?

The controller only speaks to the API... it gets watch events from some apiserver via some etcd server, but there's no guarantee the list/deletecollection calls it makes go to the same apiserver, or the same etcd server

Comment (derekwaynecarr) [inline pkg/controller/namespace/namespace_controller.go]: can you move this to a constant like "namespaceDeletionGracePeriod"

Comment (liggitt) [inline pkg/controller/namespace/namespace_controller.go]: done

**Comment by calebamiles** [comment]
adding a release blocker label to this PR as it may possibly fix https://github.com/kubernetes/kubernetes/issues/36891 which is a 1.5 release blocker.


**cc: @saad-ali, @dims**

File: pkg/controller/namespace/namespace_controller.go (hunk 1 above)
Comment (liggitt) [inline pkg/controller/namespace/namespace_controller.go]: I'm not sure #22007 helps in a case where you have namespaces stored in one etcd, and namespaced objects stored in a different etcd served by different API servers. There's not necessarily any coherence between the namespace's resourceVersion and the resourceVersion of an object in that namespace

**Comment by dims** [comment]
@k8s-bot test this

**Comment by dims** [comment]
@k8s-bot test this issue: #IGNORE

File: pkg/controller/namespace/namespace_controller.go [hunk 2]
```diff
@@ -132,7 +142,9 @@ func (nm *NamespaceController) enqueueNamespace(obj interface{}) {
 		glog.Errorf("Couldn't get key for object %+v: %v", obj, err)
 		return
 	}
-	nm.queue.Add(key)
+	// delay processing namespace events to allow HA api servers to observe namespace deletion,
+	// and HA etcd servers to observe last minute object creations inside the namespace
+	nm.queue.AddAfter(key, namespaceDeletionGracePeriod)
```
Comment (wanghaoran1988) [inline pkg/controller/namespace/namespace_controller.go]: Does this mean that I create a namespace, we cannot create resource before the `namespaceDeletionGracePeriod` ?

Comment (liggitt) [inline pkg/controller/namespace/namespace_controller.go]: No. this controller only participates in namespace deletion and cleanup

**Comment by liggitt** [comment]
@k8s-bot test this

**Comment by liggitt** [comment]
@k8s-bot unit test this

**Comment by dims** [comment]
@k8s-bot unit test this

**Comment by dims** [comment]
@k8s-bot kops aws e2e test this

**Comment by erictune** [comment]
TBR @lavalamp

**Comment by rohantmp** [comment]
Is there any way this can be changed or a flag added to disable this for one node clusters running e2e?

**Pull request merged**